src/main/scala/viper/server/vsi/Cache.scala

```diff
@@ -0,0 +1,204 @@
+    *
+    * Additionally, the method returns previously cached verification result for members of the
+    * ast that hit the cache.
+    * */
+  def apply(
+        file_key: String,
+        input_prog: AST): (AST, List[CacheEntry]) = {
+
+    val cache_entries: ListBuffer[CacheEntry] = ListBuffer()
+    val concerningsToCache: ListBuffer[Concerning] = ListBuffer()
+    val concerningsToVerify: ListBuffer[Concerning] = ListBuffer()
+
+    //read errors from cache
+    val cachable_members = input_prog.decompose()
+    cachable_members.foreach((c: Concerning) => {
+      val dependencies = c.getDependencies(input_prog)
+      get(file_key, c, dependencies) match {
+        case Some(matched_entry) =>
+          concerningsToCache += c.transform
+          cache_entries += matched_entry
+        case None =>
+          //Nothing in cache, request verification
+          concerningsToVerify += c
+      }
+    })
+    val all_concernings: List[Concerning] = concerningsToCache.toList ++ concerningsToVerify.toList
+    val output_prog: AST = input_prog.compose(all_concernings)
+    (output_prog, cache_entries.toList)
+  }
+
+  /** Utility function to retrieve entries for single members.
+    * */
+  final def get(
+        file_key: String,
+        key: Concerning,
+        dependencies: List[Member]): Option[CacheEntry] = {
+
+    val concerning_hash = key.hashFunction
+    val dependencies_hash = dependencies.map(_.hashFunction).mkString(" ")
+    val dependency_hash = CacheHelper.buildHash(concerning_hash + dependencies_hash)
+    assert(concerning_hash != null)
+
+    for {
+      fileCache <- _cache.get(file_key)
+      cacheEntries <- fileCache.get(concerning_hash)
+      validEntry <- cacheEntries.find(_.depencyHash == dependency_hash)
+    } yield validEntry
+
+
+  }
+
+  /** This function updates the cache's state by adding/updating entries.
+    * */
+  def update(
+        file_key: String,
+        key: Concerning,
+        dependencies: List[Member],
+        content: CacheContent): List[CacheEntry] = {
+
+    val concerning_hash = key.hashFunction
+    val dependencies_hash = dependencies.map(_.hashFunction).mkString(" ")
+    val dependency_hash = CacheHelper.buildHash(concerning_hash + dependencies_hash)
+    val cacheEntry: CacheEntry = new CacheEntry(key, content, dependency_hash)
+
+    assert(concerning_hash != null)
+
+    _cache.get(file_key) match {
+      case Some(fileCache) =>
+        val existing_entries = fileCache.getOrElse(concerning_hash, Nil)
+        val updated_cacheEntries = cacheEntry :: existing_entries
+
+        fileCache(concerning_hash) = updated_cacheEntries
+        updated_cacheEntries
+      case None =>
+        //if file not in cache yet, create new map entry for it. Recall the function.
+        _cache += (file_key -> collection.mutable.Map[String, List[CacheEntry]]())
+        update(file_key, key, dependencies, content)
+    }
+  }
+
+  def forgetFile(file_key: String): Option[String] = {
```

Comment:
Please add a comment for each method to describe the intended semantics of this method.